operator/internal/manifests/storage/schema_test.go

```diff
@@ -0,0 +1,145 @@
+package storage
+
+import (
+	"testing"
+	"time"
+
+	lokiv1beta1 "github.com/grafana/loki/operator/api/v1beta1"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpdateSchema_AddNewSchema(t *testing.T) {
+	expected := []Schema{
+		{
+			From:    time.Now().Format("2006-01-02"),
+			Version: lokiv1beta1.ObjectStorageSchemaV12,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV12, []Schema{})
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_AppendNewSchema(t *testing.T) {
+	expected := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+		{
+			From:    time.Now().AddDate(0, 0, DaysToUpdate).Format("2006-01-02"),
+			Version: lokiv1beta1.ObjectStorageSchemaV12,
+		},
+	}
+
+	test := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV12, test)
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_IgnoreSameSchema(t *testing.T) {
+	expected := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+	}
+
+	test := []Schema{
+		{
+			From:    "2006-01-02",
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV11, test)
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_ModifyUnappliedSchema(t *testing.T) {
+	var testVersion lokiv1beta1.ObjectStorageSchemaVersion = "v10"
+
+	from := time.Now().AddDate(0, 0, DaysToUpdate).Format("2006-01-02")
+
+	expected := []Schema{
+		{
+			From:    time.Now().Format("2006-01-02"),
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+		{
+			From:    from,
+			Version: lokiv1beta1.ObjectStorageSchemaV12,
+		},
+	}
+
+	test := []Schema{
+		{
+			From:    time.Now().Format("2006-01-02"),
+			Version: lokiv1beta1.ObjectStorageSchemaV11,
+		},
+		{
+			From:    from,
+			Version: testVersion,
+		},
+	}
+
+	actual := UpdateSchemas(lokiv1beta1.ObjectStorageSchemaV12, test)
+
+	require.Len(t, actual, len(expected))
+	for i := range expected {
+		require.Equal(t, expected[i].From, actual[i].From)
+		require.Equal(t, expected[i].Version, actual[i].Version)
+	}
+}
+
+func TestUpdateSchema_ModifyUnappliedSchema_ReduceSameVersion(t *testing.T) {
```

Comment:
This test assumes that in the `test` schema slice we have two entries where the `v11` has a later `From` date than `v12`. This case should not be possible as per meaning somehow the user tricked the system to allow downgrades. Can you elaborate on the purpose?

Reply:
Actually, this can happen. Before this PR, we have the `v11` schema that takes effect from "2020-10-01". In the same way as above, this is more functionality driven than realistic. This was testing if a user could flip back between the two versions on the same day and there was already a defined version, the earliest definition should be preferred.

However, as you mentioned, this is might be something to nit upon the API discussion. The test case above too might also be nit as well.